### PR TITLE
Error type for `from_env_ext` function

### PR DESCRIPTION
**Pull request by belovdv**
Continue implementing #51. The previous PR wasn't finished and cleaned up.

If we want to discern different errors and treat them differently, more specific return type would be more convenient.

If this's going to be merged, probably there'll be some changes from current version. If it's possible, i'd like to clean up results before merging.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -151,40 +152,80 @@ struct HelperInner {
...
-    },
+pub struct FromEnv {
+    /// Result of trying to get jobserver client from env.
+    pub client: Result<Client, FromEnvError>,
+    /// Name of the environment variable.
+    /// Empty if no relevant environment variable is found.
+    pub var_name: &'static str,
+    /// Value of the environment variable.
+    /// Empty if no relevant environment variable is found.
+    pub var_value: OsString,
```
Comment (NobodyXu) [inline src/lib.rs:165]: Why not use `Option` to represent no env variable is found and group them together as one field?
```suggestion
    /// Name and value of the environment variable.
    /// `None` if no relevant environment variable is found.
    pub var: Option<(&'static str, OsString)>,
```

File: src/lib.rs [hunk 2]
```diff
@@ -151,40 +152,80 @@ struct HelperInner {
...
+#[non_exhaustive]
+pub enum FromEnvError {
+    /// There is no environment variable that describes jobserver to inherit.
+    NoEnvVar,
+    /// Cannot parse jobserver environment variable value, incorrect format.
+    CannotParse(String),
+    /// Cannot open path or name from the jobserver environment variable value.
+    CannotOpenPath(String, io::Error),
+    /// Cannot open file descriptor from the jobserver environment variable value.
+    CannotOpenFd(std::ffi::c_int, io::Error),
```
Comment (NobodyXu) [inline src/lib.rs:179]: Is it really ok to expose unix-specific error variant?

```suggestion
    #[cfg(unix)]
    CannotOpenFd(std::os::fd::RawFd, io::Error),
```

File: src/lib.rs [hunk 3]
```diff
@@ -151,40 +152,80 @@ struct HelperInner {
...
+/// Error type for `from_env_ext` function.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum FromEnvError {
+    /// There is no environment variable that describes jobserver to inherit.
+    NoEnvVar,
+    /// Cannot parse jobserver environment variable value, incorrect format.
+    CannotParse(String),
+    /// Cannot open path or name from the jobserver environment variable value.
+    CannotOpenPath(String, io::Error),
```
Comment (NobodyXu) [inline src/lib.rs:177]:
```suggestion
    CannotOpenPath(PathBuf, io::Error),
```

**Review by NobodyXu — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 2 above)
Comment (petrochenkov) [reply src/lib.rs:179]: Code using jobserver-rs will be target-independent, most likely, so it's probably simpler to avoid cfgs.

File: src/lib.rs (hunk 1 above)
Comment (petrochenkov) [reply src/lib.rs:165]: `FromEnvError::NoEnvVar` also indicates that there's no env var and these fields can be ignored.
If there's also an `Option` they can in theory go into conflict, and you'll have to unwrap, so I guess no `Option` is simpler.

("Empty if no relevant environment ..." could easily be "Unspecified if no relevant environment ...")

File: src/lib.rs (hunk 3 above)
Comment (NobodyXu) [reply src/lib.rs:177]: Sounds reasonable , I thought about using OsString for a moment but then you would still need to convert to string in formatting.

File: src/lib.rs (hunk 1 above)
Comment (NobodyXu) [reply src/lib.rs:165]: Maybe it's better to use Result<(Client, EnvVariable), FromEnvError> and also includes EnvVariable in FromEnvError except for variant NoEmvVar and NotSupporte?

    struct EnvVariable {
        name: &'static str,
        value: OsString,
    }

File: src/lib.rs (hunk 2 above)
Comment (NobodyXu) [reply src/lib.rs:179]: Even so, I think it's better to use std::os::fd::RawFd instead of c_int

Comment (NobodyXu) [reply src/lib.rs:179]: Thanks, it's a shame that we have to use a raw integer type here and expose what I considered to be implementation detail, but I can't think of something better right now.

Comment (NobodyXu) [reply src/lib.rs:179]: @petrochenkov Could we merge CannotOpen{Path, Fd} into one and make the descriptor (path/fd) opaque?

I don't think cargo would go to the extent of inspecting which path/fd that open failed, so maybe we don't need to expose such impl detail and os specific detail?

File: src/lib.rs (hunk 3 above)
Comment (petrochenkov) [reply src/lib.rs:177]: Actually maybe it's reasonable to make it `OsString` because the env var values are `OsString`.

It's hard to parse `OsString`s without converting to `String` right now because methods like `contains` or `starts_with` are lacking, but it would be more future compatible.

File: src/lib.rs (hunk 2 above)
Comment (NobodyXu) [reply src/lib.rs:179]: > But it could probably be solved by adding a new `FromEnvError` variant too.

IMO adding a new variant is better than exposing implementation detail and let the user check it themselves.

File: src/lib.rs (hunk 1 above)
Comment (belovdv) [reply src/lib.rs:165]: In case of negative file descriptor `from_env_ext` may want to return `Client` without environment variables. Maybe it's better to make result and env independent like `(Result<Client, FromEnvError>, Option<EnvVariable>)`?

File: src/lib.rs (hunk 2 above)
Comment (belovdv) [reply src/lib.rs:179]: If fd is going to be used only for error message, would it be enough to note is it write or read descriptor, without actual number that is in environment variable value? imho negative file descriptors and things like that should be handled within this crate.

File: src/lib.rs (hunk 1 above)
Comment (NobodyXu) [reply src/lib.rs:165]: > In case of negative file descriptor `from_env_ext` may want to return `Client` without environment variables.

I don't `from_env_ext` should return Client on negative file descriptor.

Negative file descriptor means the jobserver is disabled.

In that case, the user could create a client using Client::nee

File: src/lib.rs (hunk 2 above)
Comment (petrochenkov) [reply src/lib.rs:179]: If people are not against multiplying public types, then we can go with
```rust
pub enum FromEnvError {
    NoEnvVar,
    CannotParse(CannotParseError),
    CannotOpenPath(CannotOpenPathError),
    CannotOpenFd(CannotOpenFdError),
    NotAPipe(NotAPipeError),
    Unsupported,
}
```
where `{CannotParse,CannotOpenPath,CannotOpenFd,NotAPipe}Error` are all opaque structures.

This would be future-compatible with keeping all kind of stuff inside them.

Comment (petrochenkov) [reply src/lib.rs:179]: >where `{CannotParse,CannotOpenPath,CannotOpenFd,NotAPipe}Error` are all opaque structures

They can even be made unnameable for a start.
```
mod error  {
    pub struct CannotParseError { ... }
}

pub enum FromEnvError {
    CannotParse(CannotParseError),
    ...
}
```

Comment (NobodyXu) [reply src/lib.rs:179]: If that's the case, why not make `FromEnvError` and opaque type and only expose:

```rust
pub enum FromEnvErrorKind {
    CannotParse,
    CannotOpenPath,
    CannotOpenFd,
    NotAPipe,
}

impl FromEnvError {
    pub fn kind(&self) -> FromEnvErrorKind;
}
```

Comment (petrochenkov) [reply src/lib.rs:179]: Yeah, that should be fine too.

File: src/error.rs [hunk 4]
```diff
@@ -0,0 +1,87 @@
+/// Opaque error type for `from_env_ext` function.
+#[derive(Debug)]
+pub struct FromEnvError {
+    pub(crate) inner: FromEnvErrorInner,
+}
+
+/// Public simplified representation of error type for `from_env_ext` function.
```
Comment (petrochenkov) [inline src/error.rs:7]:
```suggestion
/// Kind of an error returned from `from_env_ext` function.
```
You are writing about implementation details in documentation for users.

File: src/error.rs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    /// Cannot open file descriptor from the jobserver environment variable value.
+    CannotOpenFd,
+    /// File descriptor from the jobserver environment variable value is not a pipe.
+    NotAPipe,
+    /// Jobserver inheritance is not supported on this platform.
+    Unsupported,
+}
+
+impl FromEnvError {
+    /// Get public version of error type.
```
Comment (petrochenkov) [inline src/error.rs:26]:
```suggestion
    /// Get the error kind.
```

File: src/error.rs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum FromEnvErrorInner {
+    NoEnvVar,
+    CannotParse(String),
+    #[cfg(any(unix, windows))]
+    CannotOpenPath(String, std::io::Error),
```
Comment (petrochenkov) [inline src/error.rs:80]: `#[cfg(unix)]` on `CannotOpenFd`/`NotAPipe` allow using `RawFd`, but cfgs on this variant and `Unsupported` do not bring any benefit.

File: src/lib.rs [hunk 7]
```diff
@@ -234,10 +226,10 @@ impl Client {
     ///
     /// # Return value
     ///
+    /// `FromEnv` contains result and relevant `var_name` with `var_value`.
```
Comment (petrochenkov) [inline src/lib.rs:229]: >`var_name` with `var_value`

Outdated comment

File: src/error.rs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            FromEnvErrorInner::NotAPipe(_, Some(err)) | FromEnvErrorInner::CannotOpenFd(_, err) => {
+                Some(err)
+            }
+            _ => None,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum FromEnvErrorInner {
```
Comment (petrochenkov) [inline src/error.rs:72]:
```suggestion
enum FromEnvErrorInner {
```

File: src/error.rs [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+#[cfg(unix)]
+pub type RawFd = std::os::fd::RawFd;
+#[cfg(not(unix))]
+pub type RawFd = std::convert::Infallible;
+
+/// Error type for `from_env_ext` function.
+#[derive(Debug)]
+pub struct FromEnvError {
+    pub(crate) inner: FromEnvErrorInner,
```
Comment (petrochenkov) [inline src/error.rs:11]:
```suggestion
    inner: FromEnvErrorInner,
```

File: src/error.rs [hunk 10]
```diff
@@ -0,0 +1,79 @@
+#![allow(dead_code)]
+
+#[cfg(unix)]
+pub type RawFd = std::os::fd::RawFd;
```
Comment (petrochenkov) [inline src/error.rs:4]:
```suggestion
type RawFd = std::os::fd::RawFd;
```

File: src/error.rs [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+#[cfg(not(unix))]
+pub type RawFd = std::convert::Infallible;
```
Comment (petrochenkov) [inline src/error.rs:6]:
```suggestion
enum RawFd {} // uninhabited
```

File: src/error.rs [hunk 12]
```diff
@@ -0,0 +1,79 @@
+#![allow(dead_code)]
```
Comment (petrochenkov) [inline src/error.rs:1]: This should no longer be necessary?
(If it's still necessary it should be put on something more specific than a module.)

**Review by petrochenkov — APPROVED** [review]

**Comment by petrochenkov** [comment]
ping @alexcrichton for merging
ping @weihanglo in case you have any additional comments

(So far this PR has consensus between me, @NobodyXu and @belovdv.)

**Review by alexcrichton — COMMENTED** [review]
Sorry I haven't been following the discussion here and it seems like there's a good deal, but this to me feels like it's dialing the wrong knob per se up to 11. The `FromEnv` structure feels pretty unidiomatic as it's a struct-containing-a-`Result` which breaks `?` usage  and requires filling out fields that may not always make the most sense. I can sort of see how it's conveying more information but I would not personally settle on this particular design.

On one hand this is fine. This isn't exactly a heavily used crate by everyone in every project so it's not like something that's a bit unidiomatic will make that much of a difference. On the other hand, though, this is not code that I would want to maintain. I would prefer to improve the idioms here and figure out a different way which I'm more comfortable with which achieves the same goal. Unfortunately I don't have a ton of time or energy to go through this process. At this point it may be best for me to hand off this crate to someone else, perhaps `rust-lang`, since it's depended on by `rustc` and `cargo`. I at the same time don't want to block progress on feature work here.

**Comment by petrochenkov** [comment]
>The FromEnv structure feels pretty unidiomatic as it's a struct-containing-a-Result which breaks ? usage and requires filling out fields that may not always make the most sense.

`?` is not exactly broken because `Client::from_env_ext().client?` would still work.

But it's not too important anyway, this function is called ~once in a program, and if someone is calling `Client::from_env_ext()` instead of `Client::from_env()` he's likely going to inspect and/or log all those errors and env vars.

`FromEnv::var` is not inside the `Result` because it may exist regardless of the client being successfully constructed or not.

**Comment by alexcrichton** [comment]
I'll reiterate the second part of my comment above. I realize that this is a minor crate which is not really all that heavily used. That being said I do not want to personally maintain the idioms introduced here, nor do I necessarily have the time or energy to walk through the review process towards something I would be willing to maintain. If updating this PR requires lots of feedback from me, or if this PR would rather not be updated, then I think it's best to look for a different maintainer of this crate to transfer ownership to.

**Comment by petrochenkov** [comment]
This PR would rather not be updated.
I'll make a proposal to move this crate to rust-lang organization.

File: src/lib.rs [hunk 13]
```diff
@@ -269,27 +261,42 @@ impl Client {
...
+            .find_map(|p| p)
+        {
+            Some((env, var_os)) => (env, var_os),
+            None => return FromEnv::new_err(FromEnvErrorInner::NoEnvVar, "", Default::default()),
+        };
+
+        let var = match var_os.to_str() {
+            Some(var) => var,
+            None => {
+                let err = FromEnvErrorInner::CannotParse("not a valid UTF-8".to_string());
```
Comment (thomcc) [inline src/lib.rs:277]:
```suggestion
                let err = FromEnvErrorInner::CannotParse("not valid UTF-8".to_string());
```

**Review by thomcc — APPROVED** [review]
Looks fine with typo fixed.
